src/github.com/matrix-org/dendrite/federationapi/writers/threepid.go

```diff
@@ -80,6 +83,59 @@ func CreateInvitesFrom3PIDInvites(
 	}
 }
 
+// ExchangeThirdPartyInvite implements PUT /_matrix/federation/v1/exchange_third_party_invite/{roomID}
+func ExchangeThirdPartyInvite(
+	httpReq *http.Request,
+	request *gomatrixserverlib.FederationRequest,
+	roomID string,
+	queryAPI api.RoomserverQueryAPI,
+	cfg config.Dendrite,
+	federation *gomatrixserverlib.FederationClient,
+	producer *producers.RoomserverProducer,
+) util.JSONResponse {
+	var builder gomatrixserverlib.EventBuilder
+	if err := json.Unmarshal(request.Content(), &builder); err != nil {
+		return util.JSONResponse{
+			Code: 400,
+			JSON: jsonerror.NotJSON("The request body could not be decoded into valid JSON. " + err.Error()),
+		}
+	}
+
+	// Check that the room ID is correct.
+	if builder.RoomID != roomID {
+		return util.JSONResponse{
+			Code: 400,
+			JSON: jsonerror.BadJSON("The room ID in the request path must match the room ID in the invite event JSON"),
+		}
+	}
+
+	// Auth and build the event from what the remote server sent us
+	event, err := buildMembershipEvent(&builder, queryAPI, cfg)
+	if err == errNotInRoom {
+		return util.JSONResponse{
+			Code: 404,
+			JSON: jsonerror.NotFound("Unknown room " + roomID),
+		}
+	}
```

Comment:
We should probably check that `err != nill` as well

Reply:
Right, forgot to do it. Fixed in https://github.com/matrix-org/dendrite/pull/224/commits/40a8d732fdac61578febdda89aa52ad656bf3fb7